consensus/hotstuff/committees/consensus_committee_test.go

```diff
@@ -114,6 +114,111 @@ func TestConsensus_LeaderForView(t *testing.T) {
 	})
 }
 
+func TestRemoveOldEpochs(t *testing.T) {
+
+	identities := unittest.IdentityListFixture(10)
+	me := identities[0].NodeID
+
+	// counter for the current epoch
+	epochCounter := uint64(1)
+
+	// create mocks
+	state := new(protocolmock.ReadOnlyState)
+	snapshot := new(protocolmock.Snapshot)
+
+	epoch1 := newMockEpoch(
+		epochCounter,
+		identities,
+		1,
+		100,
+		unittest.SeedFixture(32),
+	)
+	epoch2 := newMockEpoch(
+		epochCounter+1,
+		identities,
+		101,
+		200,
+		unittest.SeedFixture(32),
+	)
+	epoch3 := newMockEpoch(
+		epochCounter+2,
+		identities,
+		201,
+		300,
+		unittest.SeedFixture(32),
+	)
+	epoch4 := newMockEpoch(
+		epochCounter+3,
+		identities,
+		301,
+		400,
+		unittest.SeedFixture(32),
+	)
+
+	state.On("Final").Return(snapshot)
+	epochs := mocks.NewEpochQuery(t, epochCounter, epoch1, epoch2, epoch3, epoch4)
+	snapshot.On("Epochs").Return(epochs)
+
+	committee, err := NewConsensusCommittee(state, me)
+	require.Nil(t, err)
+
+	// we should start with only current epoch (epoch 1) pre-computed
+	// since there is no previous epoch
+	assert.Equal(t, 1, len(committee.leaders))
+
+	// when we first request a view within epoch 2, we should not remove
+	// any old epochs since we have 2 (1 & 2)
+	_, err = committee.LeaderForView(101)
+	t.Run("first transition", func(t *testing.T) {
+		assert.Nil(t, err)
+		assert.Equal(t, 2, len(committee.leaders))
+	})
+
+	// transition so epoch 2 is current
+	epochs.Transition()
+	// request a view in epoch 3
+	_, err = committee.LeaderForView(201)
+	assert.Nil(t, err)
+
+	// when we request a view within epoch 3, we should also not remove
+	// any old epochs since we have 3 (the maximum)
+	t.Run("second transition", func(t *testing.T) {
+
+		// should have 3 epochs stored
+		assert.Equal(t, 3, len(committee.leaders))
+
+		// should have leader selection for epochs 1-3
+		_, has1 := committee.leaders[epochCounter]
+		_, has2 := committee.leaders[epochCounter+1]
+		_, has3 := committee.leaders[epochCounter+2]
+		assert.True(t, has1)
+		assert.True(t, has2)
+		assert.True(t, has3)
+	})
+
+	// transition so epoch 3 is current
+	epochs.Transition()
+	// request a view in epoch 4
+	_, err = committee.LeaderForView(301)
+	assert.Nil(t, err)
+
+	t.Run("third transition", func(t *testing.T) {
+
+		// should have 3 epochs stored (1 has been deleted)
+		assert.Equal(t, 3, len(committee.leaders))
+
+		// should have leader selection for epochs 2-4 (no 1)
+		_, has1 := committee.leaders[epochCounter]
+		_, has2 := committee.leaders[epochCounter+1]
+		_, has3 := committee.leaders[epochCounter+2]
+		_, has4 := committee.leaders[epochCounter+3]
+		assert.False(t, has1)
+		assert.True(t, has2)
+		assert.True(t, has3)
+		assert.True(t, has4)
+	})
+}
```

Comment:
would be great to have a test that use a loop to add 10 epoch transition and in the end assert there are 3 leader selections

Reply:
👍 Refactored this test to loop through 10 epochs https://github.com/onflow/flow-go/pull/191/commits/50576f3ae80a231ae2d843a82ea0e62c165b68ac